kube-scheduler: enhance performance for DRA structured parameters

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The initial implementations of the scheduler plugin and device allocation logic up to Kubernetes 1.31 was focused on correctness, not performance.

This PR contains step-by-step improvements that focus on things that stood out in a pprof analysis of the memory and CPU profiles produced by:

     ARTIFACTS=/tmp go test -memprofile=mem-prof.out -cpuprofile=cpu-prof.out  -count=6 -timeout=0 -benchtime=1ns -run=xxx -bench=BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes ./test/integration/scheduler_perf/ -args -perf-scheduling-label-filter=

Note: `ARTIFACTS` has to be set to redirect log output. Without it, `go test` and log output get interleaved so that `benchstat` only finds the first repetition.

That scenario uses moderately complex device attributes and very basic selection via a driver name selector in the device class.

Depending on the scenario, this PR  makes the average throughput up to 16 times faster:
```
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/test/integration/scheduler_perf
cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                   │            before            │                         after                         │
                                                                                   │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base                 │
PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      15.04 ± 0%                    76.68 ± 14%   +409.67% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     105.5 ± 1%                    104.8 ±  5%          ~ (p=0.457 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     95.83 ± 1%                   117.75 ±  4%    +22.88% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      79.67 ± 3%                   106.05 ±  4%    +33.11% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      27.11 ± 5%                   121.10 ±  3%   +346.70% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      84.00 ± 3%                   103.40 ±  3%    +23.10% (p=0.002 n=6)
PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      7.110 ± 6%                  121.150 ±  3%  +1603.94% (p=0.002 n=6)
geomean                                                                                                41.05                         106.2         +158.68%
```

#### Which issue(s) this PR fixes:
Related-to https://github.com/kubernetes/kubernetes/issues/113701

#### Special notes for your reviewer:

Depends on https://github.com/kubernetes/kubernetes/pull/127266.

Please review commit-by-commit. Each commit message contains a delta comparison for the individual enhancement.

Additional ideas for performance enhancements are in additional commits here: https://github.com/kubernetes/kubernetes/compare/master...pohly:kubernetes:dra-structured-performance-full

They are complex and/or didn't work out, so this is something for future experiments.

#### Note to self:

The benchstat data was added automatically with `git rebase -i origin/master`, editing each commit starting with the baseline, and then running this script:

```sh
#!/bin/bash

set -ex

run() {
    ARTIFACTS=/tmp go test -count=6 -timeout=0 -benchtime=1ns -run=xxx -bench='BenchmarkPerfScheduling/.*ResourceClaimTemplateStructured/(5000pods|empty|half|full)_[15]00nodes' ./test/integration/scheduler_perf/ -args -perf-scheduling-label-filter= | tee commit-$(git rev-parse --short HEAD).out
}

compare () {
    run

    git log -1 --pretty=format:%B >/tmp/commit
    echo >>/tmp/commit
    git log --oneline -n1
    benchstat $(ls -1 -r -t *.out | tail -2) | head -26
    benchstat before=$(ls -1 -r -t *.out | tail -2 | head -1) after=$(ls -1 -r -t *.out | tail -1) | sed -e 's/^/    /' >>/tmp/commit
    git commit --amend --no-edit -F /tmp/commit
    mv $(ls -1 -r -t *.out | tail -1) commit-$(git log --format=%f -n1).out
}

run

while git rebase --continue; do
    compare
done
```

#### Does this PR introduce a user-facing change?
```release-note
DRA: scheduling pods is up to 16x faster, depending on the scenario. Scheduling throughput depends a lot on cluster utilization. It is higher for lightly loaded clusters with free resources and gets lower when the cluster utilization increases.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

pohly committed:
DRA scheduler: avoid listing claims during Filter

The Allocate call used to call back into the claim lister for each node. This
was significant work which showed up at the top of the CPU profile. It's
okay to list only once during PreFilter because the Filter call does not change
the claim status between Allocate calls.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                        after                        │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base               │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      15.04 ± 0%                    18.06 ±  2%  +20.07% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     105.5 ± 1%                    104.7 ±  2%        ~ (p=0.485 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     95.83 ± 1%                    96.62 ±  1%        ~ (p=0.063 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      79.67 ± 3%                    83.00 ±  2%   +4.18% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      27.11 ± 5%                    32.45 ±  7%  +19.68% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      84.00 ± 3%                    95.22 ±  7%  +13.36% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      7.110 ± 6%                    9.111 ± 10%  +28.15% (p=0.002 n=6)
    geomean                                                                                                41.05                         45.86        +11.73%

pohly committed:
DRA scheduler: ResourceSlice with unique strings

Using unique strings instead of normal strings speeds up allocation with
structured parameters because maps that use those strings as key no longer need
to build hashes of the string content. However, care must be taken to call
unique.Make as little as possible because it is costly.

Pre-allocating the map of allocated devices reduces the need to grow the map
when adding devices.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                        after                         │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base                │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                     18.06 ±  2%                     33.30 ± 2%   +84.31% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                    104.7 ±  2%                     105.3 ± 2%         ~ (p=0.818 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                    96.62 ±  1%                    100.75 ± 1%    +4.28% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                     83.00 ±  2%                     90.96 ± 2%    +9.59% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                     32.45 ±  7%                     49.84 ± 4%   +53.60% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                     95.22 ±  7%                    103.80 ± 1%    +9.00% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                     9.111 ± 10%                    27.215 ± 7%  +198.69% (p=0.002 n=6)
    geomean                                                                                               45.86                           64.26        +40.12%

pohly committed:
DRA scheduler: code cleanups

Looking up the slice can be avoided by storing it when allocating a device.
The AllocationResult struct is small enough that it can be copied by value.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                       after                        │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base              │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      33.30 ± 2%                     33.95 ± 4%       ~ (p=0.288 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     105.3 ± 2%                     105.8 ± 2%       ~ (p=0.524 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     100.8 ± 1%                     100.7 ± 1%       ~ (p=0.738 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      90.96 ± 2%                     90.78 ± 1%       ~ (p=0.952 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      49.84 ± 4%                     50.51 ± 7%       ~ (p=0.485 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      103.8 ± 1%                     103.7 ± 5%       ~ (p=0.582 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      27.21 ± 7%                     28.50 ± 2%       ~ (p=0.065 n=6)
    geomean                                                                                                64.26                          64.99       +1.14%

pohly committed:
DRA scheduler: cache compiled CEL expressions

DeviceClasses and different requests are very likely to contain the same
expression string. We don't need to compile that over and over again.

To avoid hanging onto that cache longer than necessary, it's currently tied to
each PreFilter/Filter combination. It might make sense to move this up into the
scheduler plugin and thus reuse compiled expressions for different pods.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                        after                        │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base               │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      33.95 ± 4%                     36.65 ± 2%   +7.95% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     105.8 ± 2%                     106.7 ± 3%        ~ (p=0.177 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     100.7 ± 1%                     119.7 ± 3%  +18.82% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      90.78 ± 1%                    121.10 ± 4%  +33.40% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      50.51 ± 7%                     63.72 ± 3%  +26.17% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      103.7 ± 5%                     110.2 ± 2%   +6.32% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      28.50 ± 2%                     28.16 ± 5%        ~ (p=0.102 n=6)
    geomean                                                                                                64.99                          73.15       +12.56%

pohly committed:
DRA scheduler: populate set of allocated devices only once

The devices which are allocated before starting the allocation always remain
allocated. They can be stored once in a set, then each Filter call for the
different nodes can reuse that set instead of allocating it anew for each node.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                        after                         │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base                │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      36.89 ± 2%                     54.70 ± 6%   +48.26% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     105.7 ± 5%                     106.4 ± 4%         ~ (p=0.970 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     117.8 ± 3%                     120.0 ± 4%         ~ (p=0.134 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      119.5 ± 4%                     112.5 ± 4%    -5.86% (p=0.009 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      63.22 ± 2%                     87.13 ± 4%   +37.82% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      109.5 ± 2%                     113.4 ± 2%    +3.65% (p=0.006 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      27.56 ± 5%                     65.55 ± 3%  +137.84% (p=0.002 n=6)
    geomean                                                                                                72.44                          90.81        +25.37%

pohly committed:
DRA scheduler: also pre-compute the unique ResourceSlice.NodeName

Converting a node's name once to a unique string and then converting to many
unique names is faster than memory comparisons.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                       after                        │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base              │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      36.65 ± 2%                     36.89 ± 2%       ~ (p=0.452 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     106.7 ± 3%                     105.7 ± 5%       ~ (p=0.701 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     119.7 ± 3%                     117.8 ± 3%       ~ (p=0.084 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      121.1 ± 4%                     119.5 ± 4%       ~ (p=0.297 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      63.72 ± 3%                     63.22 ± 2%       ~ (p=0.485 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      110.2 ± 2%                     109.5 ± 2%       ~ (p=0.258 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      28.16 ± 5%                     27.56 ± 5%       ~ (p=0.513 n=6)
    geomean                                                                                                73.15                          72.44       -0.98%

pohly committed:
DRA scheduler: maintain a set of allocated device IDs

Reacting to events from the informer cache (indirectly, through the assume
cache) is more efficient than repeatedly listing it's content and then
converting to IDs with unique strings.

    goos: linux
    goarch: amd64
    pkg: k8s.io/kubernetes/test/integration/scheduler_perf
    cpu: Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
                                                                                       │            before            │                        after                        │
                                                                                       │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base               │
    PerfScheduling/SchedulingWithResourceClaimTemplateStructured/5000pods_500nodes-36                      54.70 ± 6%                     76.81 ± 6%  +40.42% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_100nodes-36                     106.4 ± 4%                     105.6 ± 2%        ~ (p=0.413 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/empty_500nodes-36                     120.0 ± 4%                     118.9 ± 7%        ~ (p=0.117 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_100nodes-36                      112.5 ± 4%                     105.9 ± 4%   -5.87% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/half_500nodes-36                      87.13 ± 4%                    123.55 ± 4%  +41.80% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_100nodes-36                      113.4 ± 2%                     103.3 ± 2%   -8.95% (p=0.002 n=6)
    PerfScheduling/SteadyStateClusterResourceClaimTemplateStructured/full_500nodes-36                      65.55 ± 3%                    121.30 ± 3%  +85.05% (p=0.002 n=6)
    geomean                                                                                                90.81                          106.8       +17.57%

pohly committed:
DRA scheduler: fix match attribute names in test

FullyQualifiedNames must include a domain. The current code doesn't care, but
once it does, the tests better should behave correctly.

seans3 commented:
/remove-sig api-machinery

kannon92 commented on pkg/scheduler/framework/plugins/dynamicresources/allocateddevices.go:
```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/pools.go:
```diff
@@ -21,36 +21,32 @@ import (
-	"k8s.io/apimachinery/pkg/labels"
-	resourcelisters "k8s.io/client-go/listers/resource/v1alpha3"
 	"k8s.io/component-helpers/scheduling/corev1/nodeaffinity"
+	"k8s.io/dynamic-resource-allocation/api"
```
Can you declare this to be draapi or dapi? Seeing API makes me think this is coming from somewhere else than DRA.

pohly replied:
I was wondering about this. There is other code where "api" is used for "the API we work with" instead of choosing a longer name, but I agree that a longer name may help with readability. I'm okay with "draapi".

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/cel/devicecapacityids.go:
```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
```suggestion
Copyright 2024 The Kubernetes Authors.
```

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/api/types.go:
```diff
@@ -0,0 +1,75 @@
+limitations under the License.
+*/
+
+package api
```
I really don't like the duplication of this API.

If someone updates the api in staging, it seems to be uncommon to know that these file must also be manually updated.

pohly replied:
Code generation will fail when they get out of sync by accident (deep copy).

Besides, if someone touches parts of the ResourceSlice API, they have to modify `staging/src/k8s.io/dynamic-resource-allocation/structured` and should better know how this works.

kannon92 replied:
Maybe we could call this package (api) something more fitting for generated code? Reviewing this last night, I didn't really see anything that said this was generated code.

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/api/doc.go:
```diff
@@ -0,0 +1,23 @@
+/*
```
https://github.com/kubernetes/kubernetes/pull/127277/files#r1753074927

I guess this is auto generated so this isn't so bad.

seans3 commented:
/remove-sig api-machinery

BenTheElder commented:
/remove-sig api-machinery

logicalhan reviewed (commented):
/lgtm
/approve

(for metrics test package)

dgrisonnet commented:
For the instrumentation bits

/triage accepted
/assign @logicalhan

asm582 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/celcache.go:
```diff
@@ -0,0 +1,54 @@
+
+// CELCache is an LRU cache for a compiled CEL expression.
+type CELCache struct {
+	mutex sync.Mutex
```
sync.RWMutex could allow multiple readers to access the cache

pohly replied:
This had no measurable impact. But as it makes sense at least in theory, I am including this change.

asm582 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/celcache.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (c *CELCache) get(expression string) *cel.CompilationResult {
+	c.mutex.Lock()
```
If sync.RWMutex is used then we could use `c.mutex.RLock()` and `c.mutex.RUnlock()`

asm582 commented on staging/src/k8s.io/dynamic-resource-allocation/api/uniquestring.go:
```diff
@@ -0,0 +1,41 @@
+// Returns the string that is stored in the UniqueString.
+// If the UniqueString is null, the empty string is returned.
+func (us UniqueString) String() string {
+	if us == NullUniqueString {
```
Nit: `us == UniqueString{}`

pohly replied:
For Go syntax reasons that has to be
```
if us == (UniqueString{}) {
```

I find that less readable.

asm582 commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -496,6 +516,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+	}
+
+	if err := pl.foreachPodResourceClaim(modifiedPod, nil); err != nil {
+		// This is not an unexpected error: we know that
```
Nit:  comment can be made concise are we trying to say - `This is an expected error indicating the pod is not schedulable.`

pohly replied:
This feels odd to me because "expected" implies that this is expected to occur. It doesn't always happen, it's just that if it happens, it's okay and there's no need to log or handle the error.

asm582 reviewed (commented):
Possible changes and Nit's

pohly commented:
/wg device-management

I'll rebase after https://github.com/kubernetes/kubernetes/pull/127266 is merged.

alculquicondor commented:
/assign @dom4ha

alculquicondor commented:
If this PR is about scheduling performance, why are there API changes in it?
Can you split those in a separate PR?

pohly commented:
I need to get the API changes merged first, otherwise there would be conflicts. The benchmarks also depended on a bug fix in those previous commits.

You can start reviewing commit-by-commit starting at the third commit.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/types.go:
```diff
@@ -0,0 +1,64 @@
+}
+type Device struct {
+	Name  UniqueString
+	Basic *BasicDevice
```
This conversion from the external API into a representation that is used internally is also going to be useful for https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4815-dra-partitionable-devices: that KEP introduces a new device type. It's possible to convert this `BasicDevice` into the new type here.  Then the scheduling code only needs to deal with that one.

benjaminapetersen commented:
/assign @ritazh 
for auth

pohly commented:
> for auth

Please ignore that auth change, it'll be merged via https://github.com/kubernetes/kubernetes/pull/127266. Unfortunately I need those changes here until then, otherwise the additional commits wouldn't work.

dom4ha commented on pkg/scheduler/framework/plugins/dynamicresources/allocateddevices.go:
```diff
@@ -0,0 +1,158 @@
+			// Is not considered as allocated.
+			continue
+		}
+		deviceID := structured.MakeDeviceID(result.Driver, result.Pool, result.Device)
```
Since all three params are string, it's easy to mess them up. Consider making the id from the allocation result directly.

pohly replied:
True. But `MakeDeviceID` is more than just initializing a struct, it also does the `MakeUniqueString`. There's also a "natural" order of those strings from less specific (driver) to most specific (pool). So I think the risk of mixing them up is low and the function call is justified.

dom4ha commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -526,38 +544,50 @@ func (pl *dynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
+	// Whatever is in flight also has to be checked.
+	pl.inFlightAllocations.Range(func(key, value any) bool {
+		claim := value.(*resourceapi.ResourceClaim)
+		for _, result := range claim.Status.Allocation.Devices.Results {
+			// Kubernetes 1.31 did not set this, 1.32 always does.
+			// Supporting 1.31 is not worth the additional code that
+			// would have to be written (= looking up in request) because
+			// it is extremely unlikely that there really is a result
+			// that still exists in a cluster from 1.31 where this matters.
+			if ptr.Deref(result.AdminAccess, false) {
+				// Is not considered as allocated.
+				continue
+			}
+			deviceID := structured.MakeDeviceID(result.Driver, result.Pool, result.Device)
+			logger.V(6).Info("Device is in flight for allocation", "device", deviceID, "claim", klog.KObj(claim))
+			allocated.Insert(deviceID)
 		}
```
Extracting allocated device ids from a claim is at least in three different places. Can you make it a standalone function or foreachAllocatedDevice taking lambda, especially as it has none obvious part for AdminAccess.

pohly replied:
Done. I'm struggling a bit with the naming because "admin accessed" devices are also allocated. In the end I used `foreachAllocatedDevice` and carefully explained in the function comment that this is for "exclusively allocated" devices.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -526,38 +544,50 @@ func (pl *dynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
 		// persistently.
 		//
 		// Claims are treated as "allocated" if they are in the assume cache
```
I tend to describe it as "claims are allocated if they have an allocation result" and "devices are allocated for a claim", so this comment is still correct as it talks about the claim.

pohly replied:
Ah, got it! Thanks, makes sense. I added something (just some slight wording change: `(and thus their devices)`).

pohly committed:
DRA scheduler: update some log strings

pohly committed:
DRA scheduler: refactor "allocated devices" lookup

The logic for skipping "admin access" was repeated in three different places. A
single foreachAllocatedDevices with a callback puts it into one function.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -526,38 +544,50 @@ func (pl *dynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
-		allocator, err := structured.NewAllocator(ctx, allocateClaims, &claimListerForAssumeCache{assumeCache: pl.claimAssumeCache, inFlightAllocations: &pl.inFlightAllocations}, pl.classLister, pl.sliceLister)
+		// or currently their allocation is in-flight. This does not change
+		// during filtering, so we can determine that once.
+		allocatedDevices := pl.listAllAllocatedDevices(logger)
```
Moving `listAllAllocatedDevices` into the observer doesn't work because the observer doesn't know about `pl.inFlightAllocations`. But the name clash is indeed a bit odd. I've renamed the variable here to `allAllocatedDevices`.

dom4ha replied:
I was thinking about adding inFlightAllocation to the Lister, so in fact go into direction of dra_manager.go, but I guess it will be done this way in https://github.com/kubernetes/kubernetes/pull/127904 anyway.

The rename is good enough.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/allocateddevices.go:
```diff
@@ -0,0 +1,158 @@
+			continue
+		}
+		deviceID := structured.MakeDeviceID(result.Driver, result.Pool, result.Device)
+		a.logger.V(6).Info("Device was deallocated", "device", deviceID, "claim", klog.KObj(claim))
```
Changed.

pohly committed:
DRA scheduler: rename variable

"Allocated devices" are the ones which can be observed from the informer. "All
allocated devices" also includes those which are in flight and haven't been
written back to the apiserver.

pohly commented:
I had to rebase after merging the CEL cost changes (nothing major). The new changes in response to https://github.com/kubernetes/kubernetes/pull/127277#pullrequestreview-2393862726 are in separate commits at the end, except for the trivial log message change which I squashed into the original  commit.

@dom4ha can you perhaps take another look?

pohly commented:
The base PR got merged, so I rebased. I'll get to the pending comments next.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -27,44 +27,50 @@ import (
 // NewAllocator returns an allocator for a certain set of claims or an error if
 // some problem was detected which makes it impossible to allocate claims.
+//
+// The returned Allocator is stateless. It calls the listers anew for each
```
Indeed, that's a bit out-dated now. Changed to:

> The returned Allocator can be used multiple times and is thread-safe.

That is what matters for users and explains why using the same instance in parallel Filter operations is okay.

pohly committed:
DRA scheduler: update some stale comments

pohly committed:
DRA scheduler: refactor CEL compilation cache

A better place is the cel package because a) the name can become shorter
and b) it is tightly coupled with the compiler there.

Moving the compilation into the cache simplifies the callers.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:
```diff
@@ -73,6 +79,25 @@ func (a *Allocator) ClaimsToAllocate() []*resourceapi.ResourceClaim {
 	return a.claimsToAllocate
 }
 
+func (a *Allocator) compileCELExpression(expression string) cel.CompilationResult {
```
Agreed, that's better. I think a `GetOrCompile` is fine for a cache and it fits nicely into the `cel` package, so I moved it there (`structured.CELCache` -> `cel.Cache`).

I just pushed the updated branch. Please take another look, I'd like to get this merged as soon as possible because it (indirectly, through potential code conflicts) blocks [beta promotion](https://github.com/kubernetes/kubernetes/pull/127511). It's not strictly required for beta, but would be very nice to have.

pohly commented:
Rebased once more due to a minor conflict with https://github.com/kubernetes/kubernetes/pull/128275.

dom4ha reviewed (commented):
/lgtm
The performance boost will be significant, thanks!

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/cel/cache_test.go:
```diff
@@ -0,0 +1,93 @@
+	resultTrue := cache.GetOrCompile("true")
+	require.Nil(t, resultTrue.Error)
+	resultTrueAgain := cache.GetOrCompile("true")
+	if resultTrue != resultTrueAgain {
```
Interface instances can be compared in Go as if you compared the underlying values or pointers. This can lead to runtime errors when those are not comparable, but that isn't a problem here.

pohly replied:
Not a wrong answer, but not directly applicable here. The results are structs, not interfaces. Structs that contain only comparable fields are equal if all fields are equal. *That* applies here. Because some of those fields are pointers, we get unique structs for each compilation. Some other fields are interfaces, so my previous answer wasn't completely besides the point. :grin:

dom4ha replied:
That was the only option I considered, but it's completely not obvious. Consider adding a comment.

pohly replied:
[Done](https://github.com/kubernetes/kubernetes/compare/cdb3b2400f371f9b875abfa4f2e9cc7e9b66bd3c..e2edeaafd6dfce1f2ad036562e97c09dc21332bf).

There goes my LGTM label... :sweat_smile:

pohly commented:
/assign @alculquicondor @Huang-Wei 

For approval.

Once this is merged, then I think it'll be okay to rebase and merge https://github.com/kubernetes/kubernetes/pull/127904.

cc @towca

pohly commented:
Rebased because of a very minor conflict with https://github.com/kubernetes/kubernetes/pull/128399.

dom4ha commented:
/lgtm

pohly commented:
/retest

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -526,38 +543,40 @@ func (pl *dynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
+		// or currently their allocation is in-flight. This does not change
+		// during filtering, so we can determine that once.
+		allAllocatedDevices := pl.listAllAllocatedDevices(logger)
+		slices, err := pl.sliceLister.List(labels.Everything())
```
The pod would just fail to start if it ended up in a node that no longer satisfies the requirements, right?

pohly replied:
Yes. We keep discussing how to "bounce it back to the scheduler" because there are other, non-DRA cases where that would be useful, but nothing specific has come out of that yet.

The pod can recover if the device that it needs comes back.

alculquicondor commented on staging/src/k8s.io/dynamic-resource-allocation/api/conversion.go:
why are there still API-like changes in this PR?

pohly replied:
I'm calling the package "api" because it mirrors the `k8s.io/api/resource` package, but it's not part of the Kubernetes server-side API. This conversion and usage of the type is entirely in the client.

alculquicondor replied:
So this is also improving performance for DRA drivers?

Or is scheduler a client?

alculquicondor replied:
Please clarify in PR description and/or release note, as applicable

pohly replied:
The scheduler is the client here. It uses this indirectly through the `k8s.io/dynamic-resource-allocation/structured` allocator code. I've update the PR title.

pohly replied:
This could also be `k8s.io/dynamic-resource-allocation/internal/api`, but perhaps there are other usages for it.

alculquicondor replied:
I would just be cautious about adding methods and types that arbitrary projects can use and we don't necessarily want to maintain long term. But these changes are beyond my ownership.

alculquicondor reviewed (commented):
> Please review commit-by-commit. Each commit message contains a delta comparison for the individual enhancement.

Unfortunately, the github UI doesn't allow to comment when looking at each commit individually. Please consider splitting into multiple PRs in the future.

pohly commented:
> Unfortunately, the github UI doesn't allow to comment when looking at each commit individually. 

Huh? That works for me.

> Please consider splitting into multiple PRs in the future.

I would have had to submit one commit, get it merged, then repeat for the next one because of frequent code conflicts - much too slow :cry:

alculquicondor commented:
> Huh? That works for me.

Oh sorry, it doesn't allow me to mark a file a reviewed.

pohly commented:
> Oh sorry, it doesn't allow me to mark a file a reviewed.

I see. I don't use that feature myself, so I hadn't noticed. I suppose reviewing files makes less sense when changes to a file are (intentionally, in this case) scatted across different commits. You can still review the full diff, but you are missing the explanations for why certain changes were made - that's in the individual commit messages.

alculquicondor commented:
/approve

pacoxu commented:
/hold

`./hack/unwanted-dependencies.json`  needs an update. The CI failure should be addressed.

pohly commented:
github.com/golang/groupcache/lru needed to be replaced by k8s.io/utils/lru (see https://github.com/kubernetes/kubernetes/pull/128507, merged yesterday). [Done](https://github.com/kubernetes/kubernetes/compare/e2edeaafd6dfce1f2ad036562e97c09dc21332bf..d0d6dfce741a2d3e487ac10f69f959a0246d22fb).

Can someone re-add the LGTM (@pacoxu, @dom4ha)?

pacoxu commented:
/lgtm
/unhold

pacoxu commented:
https://github.com/kubernetes/kubernetes/blob/bc79d3ba87b8b3c4b7c68f26cdfcaa35654d96ac/hack/unwanted-dependencies.json#L159

This needs an update.
/hold


```
2024/11/05 07:01:16 Status diff:
  []uint8(
  	"""
  	... // 41 identical lines
  	      "github.com/Microsoft/hnslib",
  	      "go.etcd.io/etcd/server/v3",
- 	      "go.opencensus.io"
+ 	      "go.opencensus.io",
+ 	      "k8s.io/dynamic-resource-allocation"
  	    ],
  	    "github.com/google/gofuzz": [
  	... // 93 identical lines
  	"""
  )
2024/11/05 07:01:16 Unwanted module "github.com/golang/groupcache" marked in /home/prow/go/src/k8s.io/kubernetes/hack/unwanted-dependencies.json is referenced by new dependants:
2024/11/05 07:01:16    k8s.io/dynamic-resource-allocation
2024/11/05 07:01:16 !!! Avoid updating referencing modules to versions that reintroduce use of unwanted dependencies
exit status 1
Removing /tmp/verify-vendor.sh.knDaS4
+++ exit code: 1
+++ error: 1
[0;31mFAILED   verify-vendor.sh	57s
========================
[0;31mFAILED TESTS
========================
[0;31m/home/prow/go/src/k8s.io/kubernetes/hack/verify-vendor.sh
make: *** [Makefile:133: verify] Error 1
```

pohly commented:
> This needs an update.

Ah, right - the go.mod... [Fixed](https://github.com/kubernetes/kubernetes/compare/d0d6dfce741a2d3e487ac10f69f959a0246d22fb..7863d9a3819bfae228215d94982dab50748be56c)

> /hold

You don't need to hold when there are required jobs which fail, that alone is enough to prevent merging.

pacoxu commented:
/unhold
/lgtm


> > 
> 
> You don't need to hold when there are required jobs which fail, that alone is enough to prevent merging.

Just to avoid the cibot retesting detect.

pohly commented:
> Just to avoid the ci bot retesting detect.

But the hold label doesn't prevent retesting?

/test pull-kubernetes-kind-dra 

https://github.com/kubernetes/kubernetes/issues/128410

pacoxu commented:
> But the hold label doesn't prevent retesting?

The bot will keep trying to comment `/retest` if the PR label shows that the PR is ready.  An example is like https://github.com/kubernetes/kubernetes/pull/127797#issuecomment-2415005547.

pohly commented:
Ah, now I get it. Sorry, a bit slow today... the looming code freeze is getting to me.

/test pull-kubernetes-e2e-kind-ipv6

https://github.com/kubernetes/kubernetes/issues/128563

pohly commented:
/test pull-kubernetes-integration 

https://github.com/kubernetes/kubernetes/issues/128565
